### DISCUSSION TITLE
How to get data from sensors separately in the Lora receiver??

### DISCUSSION THREAD

**Discussion by Calamardin** [sandeepmistry/arduino-LoRa#623] [Q&A]
Hello everyone, I have a question about how to separately obtain the data sent by various sensors and then assign them to different variables. Yes, for example I put a temperature sensor and a gas sensor in the same package and send it (I include a small code fragment in the transmitter)

LoRa.print(mySensor.CO2);
    LoRa.print(",");
    LoRa.print(mySensor.TVOC);
    LoRa.print(",");
And in the receiver I get everything together with the following instructions:

void onReceive(int packetSize) {
 
  for (int i = 0; i < packetSize; i++) {
   
  Serial.print((char)LoRa.read());
   How can I get the CO2 value on the one hand, the temperature on the other, etc. to assign each value to a variable and then use it, for example, to send it via mqtt?

Thanks in advance

**Comment by Kongduino** [comment]
First, you should use a data format like Cayenne LPP. The examples in [this library](https://github.com/ElectronicCats/CayenneLPP) are for LoRaWAN, but the principle is the same, regardless of how you send the data. The [decode example](https://github.com/ElectronicCats/CayenneLPP/blob/master/examples/Decode/Decode.ino) show how to transform a Cayenne data packet into a JSON object.

**Reply by Calamardin** [reply]
thank you very much for your help. Will try as soon as I can.
regards

**Comment by miroan81** [comment]
In case anyone is having the same question send your data like this separated by comma

`21.70 , 42.39 , 0.10 , 0.05 , 0.00 , 0 , 0 , 7.44 , 22.10 , 153`

Then split the string in how much parts you need

` 
    while (LoRa.available()) {
      String LoRaData = LoRa.readString();
      Serial.print(LoRaData); 

      int pos[10];
      int count = 0;
      for (int i = 0; i < LoRaData.length(); i++) {
        if (LoRaData.charAt(i) == ',') {
          pos[count] = i;
          count++;
        }
      }
      String part1 = LoRaData.substring(0, pos[0]);
      String part2 = LoRaData.substring(pos[0] + 1, pos[1]);
      String part3 = LoRaData.substring(pos[1] + 1, pos[2]);
      String part4 = LoRaData.substring(pos[2] + 1, pos[3]);
      String part5 = LoRaData.substring(pos[3] + 1, pos[4]);
      String part6 = LoRaData.substring(pos[4] + 1, pos[5]);
      String part7 = LoRaData.substring(pos[5] + 1, pos[6]);
      String part8 = LoRaData.substring(pos[6] + 1, pos[7]);
      String part9 = LoRaData.substring(pos[7] + 1, pos[8]);
      String part10 = LoRaData.substring(pos[8] + 1);`

**Reply by Kongduino** [reply]
This would be a very wasteful way of sending data... And it wouldn't work with your example, what with the spaces around the commas... A much MUCH better way would be to pack the floats into a buffer, and send the bytes. On the other side you unpack the data.

Consider this simple code:

```c
#include <stdio.h>
#include <string.h>

int main(int argc, char** argb) {
  float a, b;
  printf("Size of float: %lu\n\n", sizeof(a));
  union twoFloats {
    struct {
      float numberA, numberB;
    };
    char bytes[8];
  };
  a = 3.14;
  b = a * 2;
  union twoFloats tf0, tf1;
  printf("Size of union: %lu\n\n", sizeof(tf0));
  tf0.numberA = a;
  tf0.numberB = b;
  printf("A = %.2f, B = %.2f\n", tf0.numberA, tf0.numberB);
  memcpy(tf1.bytes, tf0.bytes, 8);
  printf("A = %.2f, B = %.2f\n", tf1.numberA, tf1.numberB);
  return 0;
}
```
Packing and unpacking the floats is easy and doesn't require much processing – and is foolproof. I put only two here, but I could have put ten. Just make the array `bytes` big enough (4 x 10 = 40 bytes). No parsing, no tokenizing, no string-to-float conversion. ie less power consumption, which is always important in IOT, especially with LoRa-enabled devices.

But that's a moot point – as I mentioned in my answer above, CayenneLPP does it all for you. But your "solution" is definitely not it...

**Reply by miroan81** [reply]
I left spaces just for debugging purposes, it was easier to read, that has been removed. The code is working great so far for the past week or so, yeah I hat to convert string to float to make db happy but that's it, I had no other issues.
TBH I wanted to avoid JSON and adding extra libraries but I'll definitely look in to it and your code above, thank you!
